lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -442,7 +442,7 @@ def is_satisfied_by(self, requirement, candidate):
         # NOTE: pre-releases is when there are several user requirements
         # NOTE: and one of them is a pre-release that also matches a
         # NOTE: transitive dependency of another requirement.
-        allow_pre_release = self._with_pre_releases or not (
+        allow_pre_release = self._with_pre_releases or requirement.is_virtual or not (
```

Comment:
This is probably not harmful but I'm pretty sure it does not address the root cause of the bug described in https://github.com/ansible/ansible/issues/78067 and https://github.com/ansible/ansible/issues/79109.

The root cause of this bug is insufficient input validation. The whole resolution logic built around the version being a string. When it is not, the broken assumptions start causing problems. I'm pretty sure, a similar bug may happen in other places that treat the version as a string.

The proper resolution could be conversion any version inputs to strings. This should happen outside of the dependency resolver so that the resolver could keep relying on this expectation in any place where this is happening. Inputs should be validated/transformed to a canonical state in an outer abstraction layer of a system so that the internal/core modules could rely on those inputs being of a predictable type/form.

It looks like this bug is coming from this line https://github.com/ansible/ansible/blob/cb2e434/lib/ansible/galaxy/dependency_resolution/dataclasses.py#L285 — if a version key is missing, it'll be `'*'` (as in "any version"). That's a string. but if it is present, there is no validation for the actual value.
It should probably be changed to also turn `null` and empty string versions to "any" asterisk `*`:
```diff
-       req_version = collection_req.get('version', '*')
+       input_version_value = collection_req.get('version')
+       req_version = str(
+           input_version_value  # int
+           if isinstance(input_version_value, int)
+           else (input_version_value or '*')  # `None` or `''` == `*`
+       )
```

It would be a good idea to have an assertion that it's a string in the data structure initializer at https://github.com/ansible/ansible/blob/cb2e434/lib/ansible/galaxy/dependency_resolution/dataclasses.py#L558.

As a possibility of a future refactoring, it could be nice to have a data structure/class/abstraction representing a version, instead of a plain string. It'd be much easier to reason about such a dedicated data type if it's used all across the resolver code base.

Reply:
I must have only scanned https://github.com/ansible/ansible/issues/78067, because I didn't realize it was `type: dir`. You're right, it doesn't fix https://github.com/ansible/ansible/issues/78067, it only impacts virtual collections, ie git and subdirs. `subdirs` shouldn't have any version associated other than '*', so this could be tightened up a bit. If it is `type: subdirs` or `type: dir`, an integer version should probably be an error as suggested in the issue.

This does fix https://github.com/ansible/ansible/issues/79109. This prerelease logic is to determine if 1) --pre flag was provided or 2) the version is an exact match. In the case of virtual collections, 2) should always be true. Below the prerelease logic, virtual collections are handled correctly without version comparison, and True is returned.

I think comparing a stringified virtual version would introduce a few issues.

Mainly, if it's not an artifact version, comparing it like it is one creates all sorts of weird cornercases. If someone has a branch name containing `<` or even looks like a semver version because it contains `.`, that should have absolutely no bearing on prereleases - it's not a semver version even if it looks like it. It's unrelated to the artifact version that will be built and installed.

I'm not sure from_requirement_dict is the correct place to do the casting. It will break a bunch of valid-version tests (there are tests for null, float, list versions, but clearly not comprehensive for `type: git`). Maybe simplest is to give an error if a `type: git` version is not a string in from_requirement_dict.

I think we may just need to exclude scm candidate/requirements from normal version evaluation (which is attempted in most places) and use == to determine if they match.

> It should probably be changed to also turn null and empty string versions to "any" asterisk *:

I don't see the need for another 'any' placeholder, but it does seem like the '*' default is problematic for `type: git`.

It should probably be HEAD instead. What about scm requirements that have '*' explicitly though in their requirements.yml? (I'm wondering now how the dependency resolver handles incompatible scm requirements - guessing it doesn't properly). Should these resolve successfully?
```yaml
- name: file://path/to/repo
  version: tagname
  type: git
- name: file://path/to/repo
  version: "*"
  type: git
```
I'd expect these to incompatible I think:
```yaml
- name: file://path/to/repo
  version: tagname
  type: git
- name: file://path/to/repo
  version: HEAD
  type: git
```

There is a somewhat related issue where the real version is replaced with '*' here https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/collection/concrete_artifact_manager.py#L316 and it's 
 helping cause #79168. The dep resolver doesn't consider virtual collections and their dependencies 'user-requested' when they are.

Much more complicated than I thought at a glance, guessing there's some invasive re-working of the scm stuff to be done here. :exploding_head: Thanks for being a sounding board, really appreciate the review.

Reply:
These are interesting observations. It does indeed sound like this needs a refactoring with better test coverage.

Reply:
After thinking more about this, I think the simple solution is just
* enforce str versions for scm requirements earlier
* add virtual dependencies to direct requirements to fix the pre-release logic

Leaving this for another PR, but since I was wondering about comparing virtual versions above, I'm thinking it might make more sense to compare `type: dir` using `verify_local_collection`ish logic. If `find_matches` did the same when evaluating preinstalled candidates, virtual collections could support `--force` too instead of always reinstalling.